Enterprise-CMCS/managed-care-review
# MCR-3570 fix: Previous contract submissions contained incorrect rate data

## Summary
[MCR-3570](https://qmacbis.atlassian.net/browse/MCR-3570)

**Bug**
 Previous contract submissions would contain incorrect rate data.

**Cause**
`contractWithHistoryToDomainModel` builds the contract revision history for a contract. Each contract revision record from the **database** will include rate revisions after the contract revision has been submitted, but `contractWithHistoryToDomainModel` did not account for these rate revisions. So they would get added to the contract revision history and remove the previous version of that rate revision. This results in the history not displaying the correct rate revision when the contract was submitted.

**Reproduction steps**
- Create and submit a contract and rate submission.
- Unlock submission and edit rate.
- Resubmit the submission.
- View the previous submission; it will have the updates you made.

**Fix**
- The fix was to ignore all rate revisions that were submitted after the contract revision submission.
- There is also a fix to preserve rate order contract revision. This was done by adding some `rate` data to the `RateRevisionType`. Then, in `ratesRevisionsToDomainModel` function, we sort by `rate.createdAt` instead of the revision createdAt.

**Additional Work**
`submitHealthPlanPackage.ts` was using the first revision of the contract history for submission instead of `draftRevision`. Changed to use the `draftRevision`

#### Test cases covered
- fixed a few tests that were not passing the strict rate id checks
-  add `submitHealthPlanPackage` test that checks revision data for rates is correct after unlock, resubmit when editing rates in place and adding new rates
- add `fetchRate` tests that check that revisions data is correct
